### PR TITLE
Iotssl 1293 add ca suppress feature

### PR DESCRIPTION
**Pull request by yanesca**
According to RFC5246 the server can indicate the known Certificate
    Authorities or can constrain the aurhorisation space by sending a
    certificate list. This part of the message is optional and if omitted,
    the client may send any certificate in the response.
    
    The previous behaviour of mbed TLS was to always send the name of all the
    CAs that are configured as root CAs. In certain cases this might cause
    usability and privacy issues for example:
    - If the list of the CA names is longer than the peers input buffer then
      the handshake will fail
    - If the configured CAs belong to third parties, this message gives away
      information on the relations to these third parties
    
    Therefore we introduce an option to suppress the CA list in the
    Certificate Request message.
    
    Providing this feature as a runtime option comes with a little cost in
    code size and advantages in maintenance and flexibility.

### CODE REVIEW

File: library/ssl_srv.c [hunk 1]
```diff
@@ -2588,36 +2588,45 @@ static int ssl_write_certificate_request( mbedtls_ssl_context *ssl )
...
-    if( ssl->handshake->sni_ca_chain != NULL )
-        crt = ssl->handshake->sni_ca_chain;
-    else
-#endif
-        crt = ssl->conf->ca_chain;
 
     total_dn_size = 0;
-    while( crt != NULL && crt->version != 0 )
+
+#if defined(MBEDTLS_SSL_SRV_C)
```
Comment (mpg) [inline library/ssl_srv.c]: Since we're in `ssl_srv.c` I think this is redundant (same for the one on the closing brace obviously).

File: include/mbedtls/ssl.h [hunk 2]
```diff
@@ -185,6 +185,9 @@
 #define MBEDTLS_SSL_PRESET_DEFAULT              0
 #define MBEDTLS_SSL_PRESET_SUITEB               2
 
+#define MDEDTLS_SSL_SUPPRESS_CA_LIST_ENABLED     1
```
Comment (gilles-peskine-arm) [inline include/mbedtls/ssl.h]: Please avoid double negatives. Make the constants e.g. `MBEDTLS_SSL_SEND_CA_LIST` and `MBEDTLS_SSL_OMIT_CA_LIST`. Or a boolean “send CA list?” if defaulting to a nonzero representation is acceptable for binary compatibility.

File: programs/ssl/ssl_server2.c [hunk 3]
```diff
@@ -124,6 +124,7 @@ int main( void )
 #define DFL_MAX_VERSION         -1
 #define DFL_ARC4                -1
 #define DFL_AUTH_MODE           -1
+#define DFL_SUPPRESS_CA         MDEDTLS_SSL_SUPPRESS_CA_LIST_DISABLED
```
Comment (gilles-peskine-arm) [inline programs/ssl/ssl_server2.c]: Don't be unafraid of double negatives, again.

File: tests/ssl-opt.sh [hunk 4]
```diff
@@ -1893,6 +1893,22 @@ run_test    "Authentication: client no cert, ssl3" \
...
+            0 \
+            -s "requested DN"
+
+run_test    "Authentication: CA suppression enabled" \
+            "$P_SRV debug_level=3 auth_mode=required suppress_ca=1" \
+            "$P_CLI crt_file=data_files/server6.crt \
+             key_file=data_files/server6.key" \
+            0 \
+            -S "requested DN"
+
```
Comment (gilles-peskine-arm) [inline tests/ssl-opt.sh]: These two test cases test the server behavior with a cooperating client, we should also test the client behavior (verify that it doesn't send a non-advertised certificate, and that it will send a certificate if the list from the server is empty), and the server behavor with a non-cooperating client (client that sends a certificate with a non-advertised CA).

File: include/mbedtls/ssl.h (hunk 2 above)
Comment (mpg) [inline include/mbedtls/ssl.h]: Historical note: in the past we tended to favor a zero representation of the default as `memset(0)` was sort-of the official way to initialize structures. Now that `mbedtls_foo_init()` is generalized, I don't think there is a reason anymore to favor zero representation for the default value.

File: tests/ssl-opt.sh (hunk 4 above)
Comment (mpg) [inline tests/ssl-opt.sh]: IIRC currently a client can only have 0 or 1 certificate provisioned, not a longer list, so our client implementation currently just ignores the CA list and sends its certificate iff it has one. So, while I agree about adding a test that the client sends a cert when the server CA list is empty, I think the other test will show that our client indeed can send a cert for non-advertised CA.

The RFC doesn't seem to explicitly state what a client is supposed to do when it only has a certificate that doesn't match the non-empty list. But I remember discussions on the TLS mailing list about similar issues (no certificate matching signature_algorithms) and it seemed to me there was a consensus that sending the cert anyway was OK: the peer has to be prepared for than anyway, and is then free to react however it likes (accept the cert anyway (unlikely), continue the handshake considering the peer as _not_ authenticated, abort the handshake).

While the last test (client sends a non-trusted cert) is not about new behavior introduced by this PR, I'm all in favor of taking that occasion to include it if we didn't test that case already (I think/hope we do, but let's make sure).

File: programs/ssl/ssl_server2.c (hunk 3 above)
Comment (yanesca) [inline programs/ssl/ssl_server2.c]: Oh, I see what you did there :)

**Comment by yanesca** [comment]
Thank you both for the fast review and the valuable feedback!

I renamed the variables and the macros to use the expression `CERT_REQ` instead of `SUPPRESS`. This way we can avoid the double negative.

Also I have added two tests:
- I have checked and the tests we already have are only testing the case when the client certificate is invalid and not the case when the issuer is not trusted, so I added a test for this case
- Also I added a test for the case when the CA list is suppressed and the client sends an untrusted certificate

File: ChangeLog [hunk 5]
```diff
@@ -1,5 +1,12 @@
 mbed TLS ChangeLog (Sorted per branch, date)
 
+= mbed TLS 2.x.x branch released xxxx-xx-xx
+
+Feature
+   * Add a new configuration option to 'mbedtls_ssl_config' to enable
+     suppressing the CA list in Certificate Request messages. The default
+     behaviour has not changed. Namely every configured CAs name is included.
```
Comment (gilles-peskine-arm) [inline ChangeLog]: nitpick: “… has not changed<strong>, n</strong>amely every …”

File: include/mbedtls/ssl.h [hunk 6]
```diff
@@ -2031,6 +2038,19 @@ void mbedtls_ssl_conf_extended_master_secret( mbedtls_ssl_config *conf, char ems
 void mbedtls_ssl_conf_arc4_support( mbedtls_ssl_config *conf, char arc4 );
 #endif /* MBEDTLS_ARC4_C */
 
+#if defined(MBEDTLS_SSL_SRV_C)
+/**
+ * \brief          Disable or enable CA list suppression
```
Comment (gilles-peskine-arm) [inline include/mbedtls/ssl.h]: Thanks for changing the name of the constants, please update the documentation to match (e.g. here “Whether to send a list of acceptable CAs in CertificateRequest messages (default: do send)”).

File: tests/ssl-opt.sh [hunk 7]
```diff
@@ -1893,6 +1910,34 @@ run_test    "Authentication: client no cert, ssl3" \
             -C "! mbedtls_ssl_handshake returned" \
             -S "X509 - Certificate verification failed"
 
+# Tests for CA list suppression
+
+run_test    "Authentication: CA suppression disabled (default)" \
```
Comment (gilles-peskine-arm) [inline tests/ssl-opt.sh]: Please update the test descriptions (e.g. “Authentication: send CA list in CertificateRequest (default)”, “Authentication: suppress CA list in CertificateRequest” or “Authentication: do not send CA lis in CertificateRequest”).

**Comment by sbutcher-arm** [comment]
Reviewed, approved, tested and merged.
